File: ebook/src/main/java/org/ebookdroid/droids/EpubContext.java
package org.ebookdroid.droids;

import org.ebookdroid.core.codec.CodecDocument;
import org.ebookdroid.droids.mupdf.codec.MuPdfDocument;
import org.ebookdroid.droids.mupdf.codec.PdfContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.util.Map;

import br.com.ebook.Config;
import br.com.ebook.foobnix.ext.CacheZipUtils;
import br.com.ebook.foobnix.ext.EpubExtractor;
import br.com.ebook.foobnix.pdf.info.ExtUtils;
import br.com.ebook.foobnix.pdf.info.JsonHelper;
import br.com.ebook.foobnix.pdf.info.model.BookCSS;
import br.com.ebook.foobnix.sys.TempHolder;

public class EpubContext extends PdfContext {

    private static final Logger LOGGER = LoggerFactory.getLogger(EpubContext.class);

    private static final String TAG = "EpubContext";
    File cacheFile;

    @Override
    public File getCacheFileName(String fileNameOriginal) {
        if (Config.SHOW_LOG)
            LOGGER.info(TAG + " | getCacheFileName: {} - {}", fileNameOriginal, BookCSS.get().hypenLang);
        cacheFile = new File(CacheZipUtils.CACHE_BOOK_DIR, (fileNameOriginal + BookCSS.get().isAutoHypens + BookCSS.get().hypenLang).hashCode() + ".epub");
        return cacheFile;
    }

    @Override
    public CodecDocument openDocumentInner(final String fileName, String password) {
        if (Config.SHOW_LOG)
            LOGGER.info(TAG + ": {}", fileName);

        if (BookCSS.get().isAutoHypens && !cacheFile.isFile())
            EpubExtractor.proccessHypens(fileName, cacheFile.getPath());

        if (TempHolder.get().loadingCancelled) {
            removeTempFiles();
            return null;
        }

        String bookPath = BookCSS.get().isAutoHypens ? cacheFile.getPath() : fileName;
        final MuPdfDocument muPdfDocument = new MuPdfDocument(this, MuPdfDocument.FORMAT_PDF, fileName, password);

        final File jsonFile = new File(cacheFile.getParentFile(), ExtUtils.getFileNameWithoutExt(cacheFile.getName()) + ".json");
        if (jsonFile.isFile()) {
            muPdfDocument.setFootNotes(JsonHelper.fileToMap(jsonFile));
            if (Config.SHOW_LOG)
                LOGGER.info("Load notes from file: {}", jsonFile);
        }

        new Thread() {
            @Override
            public void run() {
                Map<String, String> notes = null;
                try {
                    muPdfDocument.setMediaAttachment(EpubExtractor.getAttachments(fileName));
                    if (!jsonFile.isFile()) {
                        notes = EpubExtractor.get().getFooterNotes(fileName);
                        muPdfDocument.setFootNotes(notes);

                        JsonHelper.mapToFile(jsonFile, notes);
                        if (Config.SHOW_LOG)
                            LOGGER.info("save notes to file: {}", jsonFile);
                    }
                    removeTempFiles();
                } catch (Exception e) {
                    LOGGER.error("Error open document inner: {}", e.getMessage(), e);
                }
            };
        }.start();

        return muPdfDocument;
    }

}

File: ebook/src/main/java/org/ebookdroid/droids/EpubContext.kt
package org.ebookdroid.droids

import br.com.ebook.extractor.EpubBookExtractor
import br.com.ebook.foobnix.ext.CacheZipUtils
import br.com.ebook.foobnix.pdf.info.ExtUtils
import br.com.ebook.foobnix.pdf.info.JsonHelper
import br.com.ebook.foobnix.pdf.info.model.BookCSS
import br.com.ebook.foobnix.sys.TempHolder
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import org.ebookdroid.core.codec.CodecDocument
import org.ebookdroid.droids.mupdf.codec.MuPdfDocument
import org.ebookdroid.droids.mupdf.codec.PdfContext
import org.slf4j.LoggerFactory
import java.io.File

class EpubContext : PdfContext() {
    private val LOGGER = LoggerFactory.getLogger(EpubContext::class.java)
    private var cacheFile: File? = null
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    override fun freeContext() {
        scope.cancel()
        super.freeContext()
    }

    override fun getCacheFileName(fileNameOriginal: String): File {
        val hash = (fileNameOriginal + BookCSS.get().isAutoHypens + BookCSS.get().hypenLang).hashCode()
        val file = File(CacheZipUtils.CACHE_BOOK_DIR, "$hash.epub")
        cacheFile = file
        return file
    }

    override fun openDocumentInner(fileName: String, password: String): CodecDocument? {
        LOGGER.info("EpubContext openDocumentInner: {}", fileName)

        val cache = cacheFile ?: getCacheFileName(fileName)

        if (!cache.isFile) {
            EpubBookExtractor.preprocessEpub(fileName, cache.path)
        }

        if (TempHolder.get().loadingCancelled) {
            removeTempFiles()
            return null
        }

        val bookPath = cache.path
        val muPdfDocument = MuPdfDocument(this, MuPdfDocument.FORMAT_PDF, bookPath, password)

        val jsonFile = File(cache.parentFile, ExtUtils.getFileNameWithoutExt(cache.name) + ".json")
        if (jsonFile.isFile) {
            muPdfDocument.setFootNotes(JsonHelper.fileToMap(jsonFile))
            LOGGER.info("Loaded notes from file: {}", jsonFile)
        }

        scope.launch {
            try {
                muPdfDocument.setMediaAttachment(EpubBookExtractor.getAttachments(fileName))
                if (!jsonFile.isFile) {
                    val notesResult = EpubBookExtractor.extractFooterNotes(fileName)
                    val notes = notesResult.getOrNull()
                    if (notes != null) {
                        muPdfDocument.setFootNotes(notes)
                        JsonHelper.mapToFile(jsonFile, notes)
                        LOGGER.info("Saved notes to file: {}", jsonFile)
                    }
                }
                removeTempFiles()
            } catch (e: Exception) {
                LOGGER.error("Error loading Epub media/notes in Coroutine: {}", e.message, e)
            }
        }

        return muPdfDocument
    }
}
